src/input_parsers/OnnxParser.cpp

```diff
@@ -0,0 +1,1265 @@
+    {
+        int size = dim.dim_value();
+        // This is copied from the Python code, but it feels like an
+        // error should be thrown if the size is non-positive.
+        int adjustedSize = size > 0 ? size : 1;
+        result.append(adjustedSize);
+    }
+    return result;
+}
+
+TensorShape shapeOfConstant(onnx::TensorProto &constant)
+{
+    TensorShape result;
+    for (auto dim : constant.dims())
+    {
+        result.append(dim);
+    }
+    return result;
+}
+
+int tensorSize(TensorShape shape)
+{
+    int size = 1;
+    for (int dimSize : shape)
+    {
+        size *= dimSize;
+    }
+    return size;
+}
+
+TensorShape getBroadcastShape(TensorShape shape1, TensorShape shape2)
+{
+    TensorShape output;
+    auto it1 = shape1.rbegin();
+    auto it2 = shape2.rbegin();
+    while (it1 != shape1.rend() || it2 != shape2.rend())
+    {
+        int d1 = it1 == shape1.rend() ? 1 : *(it1++);
+        int d2 = it2 == shape2.rend() ? 1 : *(it2++);
+
+        if (d1 != d2)
+        {
+            String errorMessage = Stringf("Mismatch in tensor dimensions found while broadcasting (%d v %d)", d1, d2);
+            throw MarabouError(MarabouError::ONNX_PARSE_ERROR, errorMessage.ascii());
+        }
+
+        output.insertHead(d1);
+    }
+    return output;
+}
+
+/**
+ * @brief Fills in the variables (values 0 and -1) in the new shape template.
+ * See https://github.com/onnx/onnx/blob/main/docs/Operators.md#Reshape for a
+ * description of the variables.
+ *
+ * @param oldShape the previous shape of the tensor
+ * @param newShapeTemplate the template for the new shape of the tensor
+ * @param allowZero whether or not zero-valued dimensions are allowed. If they
+ * are not then all zeroes are replaced with the corresponding value in the
+ * old shape.
+ * @return
+ */
+TensorShape instantiateReshapeTemplate(TensorShape oldShape, TensorShape newShapeTemplate, bool allowZero)
+{
+    TensorShape newShape;
+    int inferredIndex = -1;
+
+    for (uint i = 0; i < newShapeTemplate.size(); i++)
+    {
+        int dimSize = newShapeTemplate[i];
+        if (dimSize == -1)
+        {
+            // Then this dimension should be inferred.
+            // Temporarily set the dimSize to be 1 so that we
+            // can use the tensorSize function to compute the
+            // size so far.
+            inferredIndex = i;
+            dimSize = 1;
+        }
+        else if (dimSize == 0)
+        {
+            if (!allowZero)
+            {
+                if (i < oldShape.size())
+                {
+                    dimSize = oldShape[i];
+                }
+            }
+        }
+        newShape.append(dimSize);
+    }
+
+    if (inferredIndex != -1)
+    {
+        newShape[inferredIndex] = tensorSize(oldShape) / tensorSize(newShape);
+    }
+    return newShape;
+}
+
+void checkEndianness()
+{
+    int num = 1;
+    bool systemIsLittleEndian = *(char *)&num == 1;
+    if (! systemIsLittleEndian)
```

Comment:
The ! operator should not have a space sparating it from the argument. 

On a general note, we use a space inside brackets on each side (except for type casting). So that should be fixed globally throughout the PR.

Reply:
This was honestly the most painful thing I've ever done in my 15 years of programming. Both aesthetically and practically. Couldn't find any C++ code formatter that supports such a style...

Reply:
Oh, wow, thanks for doing that! Clang-format has SpacesInParentheses option that does that. I am sorry, I didn't think of providing a pointer ahead of time.